### PR TITLE
Migrate NameAllocator from JVM to common

### PR DESCRIPTION
**Pull request by ForteScarlet**
Part of https://github.com/square/kotlinpoet/issues/304, https://github.com/square/kotlinpoet/pull/1959

Migrate `NameAllocator` from JVM to common and added a new internal type `CodePoint` .

> The effects of `CodePoint` related functions are not guaranteed on non-JVM platforms.

### CODE REVIEW

File: kotlinpoet/src/nonJvmMain/kotlin/com/squareup/kotlinpoet/CodePoint.nonJvm.kt [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  // TODO How check Java identifier part use code point?
+  if (charCount() == 1) {
+    return Char(code).isJavaIdentifierPart()
+  }
+
+  return true
+}
+
+internal actual fun CodePoint.charCount(): Int {
+  return if (code >= 0x010000) 2 else 1
```
Comment (JakeWharton) [inline kotlinpoet/src/nonJvmMain/kotlin/com/squareup/kotlinpoet/CodePoint.nonJvm.kt:54]: Other function doesn't use a leading 0
```suggestion
  return if (code >= 0x10000) 2 else 1
```

**Review by JakeWharton — APPROVED** [review]

File: kotlinpoet/src/commonMain/kotlin/com/squareup/kotlinpoet/NameAllocator.kt [hunk 2]
```diff
@@ -118,7 +120,7 @@ public class NameAllocator private constructor(
    */
   @JvmOverloads public fun newName(
     suggestion: String,
-    tag: Any = UUID.randomUUID().toString(),
+    tag: Any = Random.nextULong().toString(), // TODO Since Kotlin 2.0.20, it's possible to use kotlin.uuid.Uuid
```
[the remark this answers is not in the document]
Comment (ForteScarlet) [reply kotlinpoet/src/commonMain/kotlin/com/squareup/kotlinpoet/NameAllocator.kt:123]: Indeed, it seems like switching to `kotlin.uuid.Uuid` will require some additional waiting time…

[the remark this answers is not in the document]
Comment (JakeWharton) [reply kotlinpoet/src/commonMain/kotlin/com/squareup/kotlinpoet/NameAllocator.kt:123]: Maybe do `toString(16).padStart(16, '0')` so we get a sixteen-digit hexidecimal number?

File: kotlinpoet/src/jvmMain/kotlin/com/squareup/kotlinpoet/CodePoint.jvm.kt [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  Character.isLowerCase(code)
+
+internal actual fun CodePoint.isUpperCase(): Boolean =
+  Character.isUpperCase(code)
+
+internal actual fun CodePoint.isJavaIdentifierStart(): Boolean =
+  Character.isJavaIdentifierStart(code)
+
+internal actual fun CodePoint.isJavaIdentifierPart(): Boolean =
+  Character.isJavaIdentifierPart(code)
```
[the remark this answers is not in the document]
Comment (JakeWharton) [reply kotlinpoet/src/jvmMain/kotlin/com/squareup/kotlinpoet/CodePoint.jvm.kt:33]: We can do this in a follow-up.

File: kotlinpoet/src/commonMain/kotlin/com/squareup/kotlinpoet/NameAllocator.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (ForteScarlet) [reply kotlinpoet/src/commonMain/kotlin/com/squareup/kotlinpoet/NameAllocator.kt:123]: 🤔Sounds good! I will modify it later.

File: kotlinpoet/src/wasmJsMain/kotlin/com/squareup/kotlinpoet/CodePoint.wasmJs.kt [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+  if (str.length != 1) {
+    return false
+  }
+
+  return str.first().isUpperCase()
+}
+
+@Suppress("UNUSED_PARAMETER")
+private fun jsCodePointAt(str: String, index: Int): Int =
+  js("str.codePointAt(index)")
```
Comment (Egorand) [inline kotlinpoet/src/wasmJsMain/kotlin/com/squareup/kotlinpoet/CodePoint.wasmJs.kt:50]: Interesting that values returned by `js()` in `wasmJs` don't require casting same as in `js`, does WASM have its own flavour of `js()`?

**Review by Egorand — APPROVED** [review]
Nice, thanks!

File: kotlinpoet/src/wasmJsMain/kotlin/com/squareup/kotlinpoet/CodePoint.wasmJs.kt (hunk 4 above)
Comment (ForteScarlet) [reply kotlinpoet/src/wasmJsMain/kotlin/com/squareup/kotlinpoet/CodePoint.wasmJs.kt:50]: Yes, [Kotlin/JS's `js()`](https://kotlinlang.org/docs/js-interop.html) seems to be different from [Wasm's `js()`](https://kotlinlang.org/docs/wasm-js-interop.html#kotlin-functions-with-javascript-code) .

**Pull request merged**